filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilderFactory.scala

```diff
@@ -1,27 +1,120 @@
+    container, endpoint, Option(workspaceId), Option(workspaceManagerURL))
 
   override def withOptions(options: WorkflowOptions)(implicit as: ActorSystem, ec: ExecutionContext): Future[BlobPathBuilder] = {
     Future {
-      new BlobPathBuilder(new AzureSasCredential(sasToken), container, endpoint)
+      new BlobPathBuilder(blobTokenGenerator, container, endpoint)
+    }
+  }
+}
+
+sealed trait BlobTokenGenerator {
+  def getAccessToken: String
+}
+
+object BlobTokenGenerator {
+  def createBlobTokenGenerator(container: String, endpoint: String): BlobTokenGenerator = {
+    createBlobTokenGenerator(container, endpoint, None, None)
+  }
+  def createBlobTokenGenerator(container: String, endpoint: String, workspaceId: Option[String], workspaceManagerURL: Option[String]): BlobTokenGenerator = {
+     (container: String, endpoint: String, workspaceId, workspaceManagerURL) match {
+       case (container, endpoint, None, None) =>
+         NativeBlobTokenGenerator(container, endpoint)
+       case (container, endpoint, Some(workspaceId), Some(workspaceManagerURL)) =>
+         WSMBlobTokenGenerator(container, endpoint, workspaceId, workspaceManagerURL)
+       case _ =>
+         throw new Exception("Bad! >:(")
+     }
+  }
+}
+
+case class WSMBlobTokenGenerator(container: String, endpoint: String, workspaceId: String, workspaceManagerURL: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
+    throw new NotImplementedError
+  }
+}
+
+case class NativeBlobTokenGenerator(container: String, endpoint: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
+    val storageAccountName = BlobPathBuilder.parseStorageAccount(BlobPathBuilder.parseURI(endpoint)) match {
+      case Some(storageAccountName) => storageAccountName
+      case _ => throw new Exception("Storage account could not be parsed from endpoint")
+    }
+
+    val profile = new AzureProfile(AzureEnvironment.AZURE)
+    val azureCredential = new DefaultAzureCredentialBuilder()
+      .authorityHost(profile.getEnvironment.getActiveDirectoryEndpoint)
+      .build
+    val azure = AzureResourceManager.authenticate(azureCredential, profile).withDefaultSubscription
+
+    val storageAccounts = azure.storageAccounts()
+    val storageAccount = storageAccounts
+      .list()
+      .asScala
+      .find(_.name == storageAccountName)
+
+    val storageAccountKeys = storageAccount match {
+      case Some(value) => value.getKeys().asScala.map(_.value())
+      case _ => throw new Exception("Storage Account not found")
+    }
+
+    val storageAccountKey = storageAccountKeys.headOption match {
+      case Some(value) => value
+      case _ => throw new Exception("Storage Account has no keys")
     }
+
+    val keyCredential = new StorageSharedKeyCredential(
+      storageAccountName,
+      storageAccountKey
+    )
+    val blobServiceClient = new BlobServiceClientBuilder()
```

Comment:
Another FYI, based on my investigation this may need to change to BlobContainerClientBuilder(), which would also mean we no longer need to supply the service and resource types. I was unsure if this is a change we want to make as a part of this patch, or as a follow on patch with a new issue.